libs/contrib/Data/Monoid/Exponentiation.idr

```diff
@@ -0,0 +1,83 @@
+module Data.Monoid.Exponentiation
+
+import Control.Algebra
+import Data.Nat.Views
+import Data.Num.Implementations
+import Syntax.PreorderReasoning
+
+%default total
+
+------------------------------------------------------------------------
+-- Implementations
+
+public export
+linear : Monoid a => a -> Nat -> a
+linear v Z = neutral
+linear v (S k) = v <+> linear v k
+
+public export
+modularRec : Monoid a => a -> HalfRec n -> a
+modularRec v HalfRecZ = neutral
+modularRec v (HalfRecEven _ acc) = let e = modularRec v acc in e <+> e
+modularRec v (HalfRecOdd _ acc) = let e = modularRec v acc in v <+> e <+> e
+
+public export
+modular : Monoid a => a -> Nat -> a
+modular v n = modularRec v (halfRec n)
+
+infixl 10 ^
```

Comment:
Is there a reason you make it *left*-associative? The common convention seems that exponentation is regarded as a right-associative operation.

Reply:
It was a purely type-directed choice. In general `2 ^ m` has type `a` which makes it an
ideal candidate for further `^ m`  action at the same type. If you make it right associative,
the nested call to `^`  uses `Num Nat` instead of `Num a`.

Reply:
Would that be a problem, if the nested call used `Num Nat`? Right-associativity seems to be the universal convention, in mathematics as well as in programming languages.

Reply:
> Would that be a problem, if the nested call used `Num Nat`? Right-associativity seems to be the (near-)universal convention, in mathematics as well as in programming languages.

My experience doesn't let me to agree with you. E.g. [wikipedia does not agree with you either](https://en.wikipedia.org/wiki/Exponentiation#Identities_and_properties)

Reply:
This wikipedia article does suggest we should make it right associative:

> Without parentheses, the conventional order of operations for serial exponentiation in superscript notation is top-down (or right-associative), not bottom-up[19][20][21][22] (or left-associative). That is,

Reply:
> > Would that be a problem, if the nested call used `Num Nat`? Right-associativity seems to be the (near-)universal convention, in mathematics as well as in programming languages.
> 
> My experience doesn't let me to agree with you. E.g. [wikipedia does not agree with you either](https://en.wikipedia.org/wiki/Exponentiation#Identities_and_properties)

This wikipedia page states: 

>  conventional order of operations for serial exponentiation in superscript notation is top-down (or right-associative

How does that not agree with me?

In any case, using right-associativity has practical benefits, since e.g.  `((x^y)^z)^w` and already be written as `x^(y*z*w)`.

Reply:
Fixed. Thanks for the review!